### ISSUE TITLE
Lint/ReturnInVoidContext should not work if there is a condition

### ISSUE THREAD

**Issue by AlexWayfer** [bbatsov/rubocop#4735]
Hello!

The new `Lint/ReturnInVoidContext` cop is good, but I found an unpleasant situation.

--------

## Expected behavior

Code:

```ruby
## Set (or delete) cookies for response
## @param key [String, Symbol] name of cookie
## @param new_value [Object, nil] value of cookie
## @example Set new value to `cat` cookie
##   cookies['cat'] = 'nice cat'
## @example Delete `cat` cookie
##   cookies['cat'] = nil
def []=(key, new_value)
  return @response.delete_cookie(key.to_s, path: '/') if new_value.nil?
  @response.set_cookie(key.to_s, value: new_value, path: '/')
end
```

`Lint/ReturnInVoidContext` should not work, I guess, for situations, where `return` used with condition (void context doesn't care about `return` anyway, does it?).

## Actual behavior

`Lint/ReturnInVoidContext` triggers and it forces you to write more code:

```ruby
def []=(key, new_value)
  if new_value.nil?
    @response.delete_cookie(key.to_s, path: '/')
  else
    @response.set_cookie(key.to_s, value: new_value, path: '/')
  end
end
```

## Steps to reproduce the problem

Write a code as in example above, run `rubocop >= 0.50`.

## RuboCop version

```
$ rubocop -V
0.50.0 (using Parser 2.4.0.0, running on ruby 2.4.2 x86_64-linux)
```

**Comment by mikegee** [comment]
This is packing too work into one line IMO:

```ruby
return @response.delete_cookie(key.to_s, path: '/') if new_value.nil?
```

Its like a regular guard clause, but doing some extra work before exiting the method by sticking it into `return`'s argument. 

I think the cop is justly prompting you to reconsider.

**Comment by AlexWayfer** [comment]
> but doing some extra work before exiting the method by sticking it into return's argument.

I'm ready to put up with this "extra work".

> I think the cop is justly prompting you to reconsider.

…and write 4 LoC instead of 2, yep. I'm not sure that getting rid of "extra work with `return`" is worth it. I doubt that it's very costly in performance.

**Comment by mikegee** [comment]
I wasn't referring to line count or performance, just clarity.

Anyways, your code is providing an argument to `return` when it will be ignored. The cop isn't wrong for pointing that out. The debate is only whether this pattern should be exempt.

**Comment by AlexWayfer** [comment]
> I wasn't referring to line count or performance, just clarity.

I referred to LoC, you referred to "extra work", what is meant "performance", I think.

> Anyways, your code is providing an argument to return when it will be ignored. The cop isn't wrong for pointing that out. The debate is only whether this pattern should be exempt.

Yes.

**Comment by Drenmi** [comment]
Ignoring the problems of the specific example, the question of whether guard clauses in a void context method definition are okay seems legitimate.

On one hand, if you have a guard clause in a method without return value, you have no indication of which execution path the code took. On the other hand, in some cases that might not matter, as long as the behaviour is the same for both paths. 🤔

**Comment by mikegee** [comment]
> guard clauses in a void context method definition

To be clear, the offense isn't because of the guard clause or the conditional, only because `return` is given an argument. This code has no offense:

```ruby
def []=(key, new_value)
  return if new_value.nil?
  @response.set_cookie(key.to_s, value: new_value, path: '/')
end
```

**Comment by Drenmi** [comment]
@mikegee: Then I agree this cop is behaving as intended. Saying `return foo if bar?` from within a void context is misleading at best. 🙂

**Comment by AlexWayfer** [comment]
> This code has no offense:

And doesn't have cookie deletion, yes, great :confused: 

> Saying return foo if bar? from within a void context is misleading at best. :slightly_smiling_face:

allows you to write less code at best * :pensive:

**Comment by Drenmi** [comment]
> And doesn't have cookie deletion, yes, great 😕

It looks like this object wants to implement a `Hash` like interface. To delete something from a hash, you use the `Hash#delete` method. So this could be an idea. Or just using an `if-else` clause over a guard clause also works.

Right now the API is a bit obscure. The method is overloaded with a special value that if you associate a key to that value, the key is deleted. There's no way to find that out without going in and reading the documentation or the code. (Unless this is a very strong convention in the project.)

It results in this strange trade-off that, in order to make the method less explicit about what it does (deleting something), it is forced to be more explicit about something it doesn't do (returning something.)

> allows you to write less code at best * 😔

Less code is not an end in itself. We could probably meta-program most applications down to a few 100 lines of code, and make it completely impenetrable in the process.

What we want is code that is readable, understandable, and clear in its intent. Often less code is easer to read and understand, so it can work as a proxy measure, but there's a balance to be had. 🙂

**Comment by AlexWayfer** [comment]
@Drenmi OK, thanks.
